File: others/tic-tac-toe.py
def evaluate(board):
    # 所有的获胜组合：3行，3列，2条对角线
    win_patterns = [
        [0, 1, 2],
        [3, 4, 5],
        [6, 7, 8],  # 行
        [0, 3, 6],
        [1, 4, 7],
        [2, 5, 8],  # 列
        [0, 4, 8],
        [2, 4, 6],  # 对角线
    ]

    for pattern in win_patterns:
        if board[pattern[0]] == board[pattern[1]] == board[pattern[2]]:
            if board[pattern[0]] == "X":
                return 10  # AI 赢
            elif board[pattern[0]] == "O":
                return -10  # 人类赢
    return 0  # 没分出胜负或平局


def is_moves_left(board):
    return " " in board


def minimax(board, depth, is_maximizing):
    score = evaluate(board)

    # 1. 递归终止条件：有人赢了，或者棋盘满了（平局）
    if score == 10:
        return score - depth  # AI 赢，步数越少越好
    if score == -10:
        return score + depth  # 人类赢，步数越多越好（尽可能拖延）
    if not is_moves_left(board):
        return 0

    # 2. 如果轮到 AI (Maximizer) 走棋
    if is_maximizing:
        best_score = -float("inf")
        for i in range(9):
            if board[i] == " ":
                board[i] = "X"  # 假设 AI 走这一步
                # 递归计算这一步的最终得分，下一轮轮到 Minimizer
                current_score = minimax(board, depth + 1, False)
                board[i] = " "  # 撤销这一步（回溯）
                best_score = max(best_score, current_score)
        return best_score

    # 3. 如果轮到人类 (Minimizer) 走棋
    else:
        best_score = float("inf")
        for i in range(9):
            if board[i] == " ":
                board[i] = "O"  # 假设人类走这一步
                # 递归计算这一步的最终得分，下一轮轮到 Maximizer
                current_score = minimax(board, depth + 1, True)
                board[i] = " "  # 撤销这一步（回溯）
                best_score = min(best_score, current_score)
        return best_score


def find_best_move(board):
    best_score = -float("inf")
    best_move = -1

    for i in range(9):
        if board[i] == " ":
            board[i] = "X"  # 尝试落子
            # 计算这步棋的价值（由于这步是 AI 走的，下一层递归应该是 False/人类的回合）
            move_val = minimax(board, 0, False)
            board[i] = " "  # 撤销落子

            # 记录产生最高分数的移动位置
            if move_val > best_score:
                best_move = i
                best_score = move_val

    return best_move


def print_board(board):
    for i in range(0, 9, 3):
        print(f" {board[i]} | {board[i+1]} | {board[i+2]} ")
        if i < 6:
            print("---+---+---")


# 初始棋盘
current_board = [" "] * 9

while is_moves_left(current_board) and evaluate(current_board) == 0:
    # 假设人类先手 (O)
    human_move = int(input("轮到你了 (输入 0-8 选择位置): "))
    if current_board[human_move] == " ":
        current_board[human_move] = "O"
    else:
        print("这个位置已经被占用了！")
        continue

    print_board(current_board)
    print("=" * 15)

    if not is_moves_left(current_board) or evaluate(current_board) != 0:
        break

    # AI 回合 (X)
    print("AI 正在思考...")
    ai_move = find_best_move(current_board)
    current_board[ai_move] = "X"
    print_board(current_board)
    print("=" * 15)

# 判断最终结果
final_score = evaluate(current_board)
if final_score > 0:
    print("AI 赢了！")
elif final_score < 0:
    print("你赢了！(这理论上是不可能的)")
else:
    print("平局！")

File: others/tic-tac-toe.cpp
#include <cstdio>
#include <iostream>
using namespace std;

// human: x
// ai: o
// empty: NULL

constexpr int N = 5;
constexpr int inf = 0x3f3f3f3f;

char board[N][N];

void init() {
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            board[i][j] = 0;
}

int evaluate() {
    /*
    since the game state space is small, we can directly check if there is a
    winner. If there is a winner, return 10 for ai and -10 for human. If there
    is no winner, return 0.

    otherwise, evaluate function should return a score based on the current
    board state.
    */
    auto ret_score = [&](int x, int y) -> int {
        return board[x][y] == 'o' ? 10 : -10;
    };

    auto &b = board;

    for (int i = 0; i < 3; i++) {
        if (b[i][0] && b[i][0] == b[i][1] && b[i][1] == b[i][2])
            return ret_score(i, 0); // row
        if (b[0][i] && b[0][i] == b[1][i] && b[1][i] == b[2][i])
            return ret_score(0, i); // column
    }
    // diagonal
    if (b[0][0] && b[0][0] == b[1][1] && b[1][1] == b[2][2])
        return ret_score(0, 0);
    if (b[0][2] && b[0][2] == b[1][1] && b[1][1] == b[2][0])
        return ret_score(0, 2);
    return 0;
}

bool is_moves_left() {
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            if (!board[i][j])
                return true;
    return false;
}

int minmax(int dep, bool is_max) {
    /*
    best outcome for ai is 10, worst outcome for ai is -10, and draw is 0.
    if ai wins, return 10 - depth (to prefer faster wins)
    if human wins, return -10 + depth (to prefer slower losses)
    return best score that can be achieved from the current board state,
    assuming both players play optimally.
    */
    int score = evaluate();
    if (score == 10)
        return score - dep;
    if (score == -10)
        return score + dep;
    if (!is_moves_left())
        return 0;

    if (is_max) { // ai's turn
        int best = -inf;
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (!board[i][j]) {
                    board[i][j] = 'o';
                    best = max(best, minmax(dep + 1, false));
                    board[i][j] = 0;
                }
            }
        }
        return best;
    } else { // human's turn
        int best = inf;
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (!board[i][j]) {
                    board[i][j] = 'x';
                    best = min(best, minmax(dep + 1, true));
                    board[i][j] = 0;
                }
            }
        }
        return best;
    }
}

tuple<int, int> find_best_move() {
    int bval = -inf, bx = -1, by = -1;
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            if (!board[i][j]) {
                board[i][j] = 'o';
                int val = minmax(0, false);
                board[i][j] = 0;
                if (val > bval) {
                    bval = val;
                    bx = i;
                    by = j;
                }
            }
        }
    }
    return {bx, by};
}

void print_board() {
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            cout << (board[i][j] ? board[i][j] : '.') << " ";
        }
        cout << endl;
    }
}

int main() {
    init();

    while (true) {
        fflush(stdout);
        print_board();
        int x, y;
        cout << "Enter your move (row and column): ";
        cin >> x >> y;
        if (board[x][y]) {
            cout << "Invalid move. Try again." << endl;
            continue;
        }
        board[x][y] = 'x';
        if (evaluate() == -10) {
            print_board();
            cout << "You win!" << endl;
            break;
        }
        if (!is_moves_left()) {
            print_board();
            cout << "It's a draw!" << endl;
            break;
        }
        auto [bx, by] = find_best_move();
        board[bx][by] = 'o';
        if (evaluate() == 10) {
            print_board();
            cout << "AI wins!" << endl;
            break;
        }
    }

    return 0;
}